### PR TITLE
Resync topology to guest cluster

### PR DESCRIPTION
**Pull request by chrisho**
**Problem:**
https://github.com/harvester/harvester/issues/3142

**Solution:**
After migrating the VM, CCM watches the VirtualMachineInstances, which will re-sync the node's topology labels of the guest cluster and ensure the guest cluster node is the same as the VM's host topology.

**Related Issue:**
https://github.com/harvester/harvester/issues/3142

**Environment:**
- At least two harvester nodes.

**Test plan:**
1. Add labels topology.kubernetes.io/zone or topology.kubernetes.io/region for per harvester node.
image
2. Create a rke2 guest cluster on Rancher.
3. Wait for the guest cluster to be completed.
4. Replace the CCM pod with the re-sync topology image.
5. Migrate the guest VM to another node.
image
6. Watch the guest cluster node's topology label to be changed.
image

### CODE REVIEW

File: pkg/cloud-controller-manager/ccm.go [hunk 1]
```diff
@@ -30,7 +36,7 @@ func newCloudProvider(reader io.Reader) (cloudprovider.Interface, error) {
 	if err != nil {
 		return nil, err
 	}
-	clientConfig, err := config.ClientConfig()
+	ClientConfig, err = config.ClientConfig()
```
Comment (yaocw2020) [inline pkg/cloud-controller-manager/ccm.go:39]: Generally, the variable name starts with a lower letter.

File: pkg/cloud-controller-manager/ccm.go [hunk 2]
```diff
@@ -39,17 +45,18 @@ func newCloudProvider(reader io.Reader) (cloudprovider.Interface, error) {
 		return nil, err
 	}
 
-	ns := rawConfig.Contexts[rawConfig.CurrentContext].Namespace
+	Namespace = rawConfig.Contexts[rawConfig.CurrentContext].Namespace
```
Comment (yaocw2020) [inline pkg/cloud-controller-manager/ccm.go:48]: ditto

File: pkg/util/constant.go [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package util
+
+const (
```
Comment (yaocw2020) [inline pkg/util/constant.go:3]: Where do we use these three const variables?

File: pkg/cloud-controller-manager/ccm.go (hunk 1 above)
Comment (chrisho) [reply pkg/cloud-controller-manager/ccm.go:39]: The variable is global and the controller uses the variable. pkg/controller/setup.go#23

File: pkg/cloud-controller-manager/ccm.go (hunk 2 above)
Comment (chrisho) [reply pkg/cloud-controller-manager/ccm.go:48]: ditto, same as ClientConfig

File: pkg/controller/virtualmachineinstance/controller.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+type Handler struct {
+	vmis       ctlv1.VirtualMachineInstanceController
+	vmiCache   ctlv1.VirtualMachineInstanceCache
+	nodeCache  ctlcorev1.NodeCache
+	restClient kubernetes.Interface
+}
+
+func (h *Handler) OnVmiChanged(_ string, vmi *kubevirtv1.VirtualMachineInstance) (*kubevirtv1.VirtualMachineInstance, error) {
+	if vmi == nil || vmi.DeletionTimestamp != nil ||
+		vmi.Annotations == nil || vmi.Status.MigrationState == nil {
```
[the remark this answers is not in the document]
Comment (chrisho) [reply pkg/controller/virtualmachineinstance/controller.go:44]: After the initial migration is completed, MigrationState will continue to exist and cannot be used as an expected value. Unless we can mark it on the VMI after resync, but the benefits are not significant.

File: go.mod [hunk 5]
```diff
@@ -3,162 +3,213 @@ module github.com/harvester/harvester-cloud-provider
...
+	github.com/rancher/system-upgrade-controller/pkg/apis v0.0.0-20210727200656-10b094e30007 // indirect
+	github.com/spf13/cobra v1.5.0 // indirect
+	go.etcd.io/etcd/api/v3 v3.5.4 // indirect
+	go.etcd.io/etcd/client/pkg/v3 v3.5.4 // indirect
+	go.etcd.io/etcd/client/v3 v3.5.4 // indirect
 	go.opentelemetry.io/contrib v0.20.0 // indirect
 	go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc v0.20.0 // indirect
 	go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp v0.20.0 // indirect
-	go.opentelemetry.io/otel v0.20.0 // indirect
+	go.opentelemetry.io/otel v1.6.3 // indirect
```
[the remark this answers is not in the document]
Comment (chrisho) [reply go.mod:174]: There are different versions of otel dependencies in the library of dependencies. If not replaced, the following problems may occur.
```
github.com/harvester/harvester-cloud-provider imports
        k8s.io/cloud-provider/options imports
        k8s.io/apiserver/pkg/server/options imports
        go.opentelemetry.io/otel/semconv: module go.opentelemetry.io/otel@latest found (v1.14.0), but does not contain package go.opentelemetry.io/otel/semconv
github.com/harvester/harvester-cloud-provider imports
        k8s.io/cloud-provider/options imports
        k8s.io/apiserver/pkg/server/options imports
        go.opentelemetry.io/otel/exporters/otlp/otlpgrpc imports
        go.opentelemetry.io/otel/exporters/otlp imports
        go.opentelemetry.io/otel/sdk/metric/controller/basic imports
        go.opentelemetry.io/otel/metric/registry: module go.opentelemetry.io/otel/metric@latest found (v0.37.0), but does not contain package go.opentelemetry.io/otel/metric/registry
```

**Comment by yaocw2020** [comment]
It would be more elegant to move the controller into CloudProvider.Initialize method. For example, 
```
func (c *CloudProvider) Initialize(clientBuilder cloudprovider.ControllerClientBuilder, stop <-chan struct{}) {
	go func() {
		if err := start.All(context.TODO(), 2, c.localCoreFactory); err != nil {
			klog.Fatal(err)
		}
		<-stop
	}()
}
```

**Comment by chrisho** [comment]
> It would be more elegant to move custom controllers into CloudProvider.Initialize method. For example,
> 
> ```
> func (c *CloudProvider) Initialize(clientBuilder cloudprovider.ControllerClientBuilder, stop <-chan struct{}) {
> 	go func() {
> 		if err := start.All(context.TODO(), 2, c.localCoreFactory); err != nil {
> 			klog.Fatal(err)
> 		}
> 		<-stop
> 	}()
> }
> ```

In next release, we could consider improving like this: https://github.com/kubernetes/cloud-provider-gcp/blob/c35e977544e40b9df0b476201acbea0ea429dd54/cmd/cloud-controller-manager/main.go#L55-L65, It is important to note that the combination with rancher framework to improve.

**Comment by chrisho** [comment]
@yaocw2020 @guangbochen ready for review, thanks.

**Comment by chrisho** [comment]
rearranged code

**Review by guangbochen — APPROVED** [review]
LGTM, and verified that the new typology label is synced correctly after the VM live migration.
image

**Review by yaocw2020 — APPROVED** [review]
LGTM, thanks

**Review by futuretea — APPROVED** [review]
LGTM for the resync logic.
Tested by follow the test plan

**Pull request merged**